enigma/src/main/java/cuchaz/enigma/translation/mapping/serde/MappingFormat.java

```diff
@@ -67,25 +69,60 @@ public MappingsReader getReader() {
 		return this.reader;
 	}
 
-	public static MappingFormat[] getReadableFormats() {
-		return new MappingFormat[] {
-				ENIGMA_DIRECTORY,
-				ENIGMA_FILE,
-				TINY_V2,
-				TINY_FILE,
-				ENIGMA_ZIP,
-				PROGUARD
-		};
-	}
+	/**
+	 * Determines the mapping format of the provided file. Checks all formats, and returns {@link #RECAF} if none match.
+	 * @param file the file to analyse
+	 * @apiNote While most assessments are based on file extension, tiny format is determined by the contents of the file.
+	 * If it contains a tiny header ("tiny[tab]2[tab]0"), the file is considered {@link #TINY_V2}. Otherwise, it goes to {@link #TINY_FILE}.
+	 * Any directory is considered to be the {@link #ENIGMA_DIRECTORY} format. Recaf and Proguard formats are not determined by file extension,
+	 * but we only check for Proguard, Recaf being the fallback.
+	 * @return the mapping format of the file.
+	 */
+	public static MappingFormat parseFromFile(Path file) {
+		if (Files.isDirectory(file)) {
+			return ENIGMA_DIRECTORY;
+		} else {
+			switch (MoreFiles.getFileExtension(file).toLowerCase()) {
+				case "zip" -> {
+					return ENIGMA_ZIP;
+				}
+				case "mapping" -> {
+					return ENIGMA_FILE;
+				}
+				case "tiny" -> {
+					// the first line of a tiny v2 file should be a header with tiny[tab]2[tab]0
+					try {
+						String contents = Files.readString(file);
+
+						if (contents.contains("tiny\t2\t0")) {
+							return TINY_V2;
+						} else {
+							return TINY_FILE;
+						}
+					} catch (IOException e) {
+						return TINY_V2;
+					}
+				}
+				case "tsrg" -> {
+					return SRG_FILE;
+				}
+				default -> {
+					// check for proguard. Recaf is the default if we don't match proguard here
+					try {
+						String contents = Files.readString(file);
+						String firstLine = contents.split("\n")[0];
+						String[] splitFirstLine = firstLine.split(" ");
+
+						if (splitFirstLine.length == 3 && splitFirstLine[1].equals("->") && splitFirstLine[2].endsWith(":")) {
+							return PROGUARD;
+						}
+					} catch (IOException e) {
+						return RECAF;
```

Comment:
Isn't this `IOException` only possible if the file can't be read?

Reply:
yes, I think that was a bad decision on my part
should this just rethrow?

Reply:
Yeah, a RuntimeException should do